Gossiper state is absent after full cluster restart with consistent topology changes enabled

This manifests as a failure in  `test_cluster_features.test_joining_old_node_fails` (dev mode):

```
    
        # Wait until the feature is considered enabled by all nodes
        cql = manager.cql
        hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
        logging.info(f"Waiting until {TEST_FEATURE_NAME} is enabled on all nodes")
>       await asyncio.gather(*(wait_for_feature(TEST_FEATURE_NAME, cql, h, time.time() + 60) for h in hosts))

test/topology/test_cluster_features.py:141: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/pylib/util.py:129: in wait_for_feature
    await wait_for(feature_is_enabled, deadline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pred = <function wait_for_feature.<locals>.feature_is_enabled at 0x7f8540516160>
deadline = 1689170433.9396355, period = 1

    async def wait_for(
            pred: Callable[[], Awaitable[Optional[T]]],
            deadline: float, period: float = 1) -> T:
        while True:
>           assert(time.time() < deadline), "Deadline exceeded, failing test."
E           AssertionError: Deadline exceeded, failing test.

test/pylib/util.py:43: AssertionError
```

The feature is not getting enabled as expected. The reason is that gossiper doesn't see that other nodes support it. 

One of the reasons is that endpoint states are not restored after restart. 

See that `loaded_endpoints` is empty:

```
INFO  2023-07-12 23:01:51,683 [shard 0] storage_service - initial_contact_nodes={127.49.130.4}, loaded_endpoints={}, loaded_peer_features=2

```

This was disabled in fd1b1279c4324bc06a2d7710edc55546e72b93a5.

This also manifests as the following warnings in logs (triggered from `gossiper::is_alive()`:
```
WARN  2023-07-12 23:02:54,868 [shard 0] gossip - unknown endpoint 127.49.130.2
WARN  2023-07-12 23:02:54,868 [shard 0] gossip - unknown endpoint 127.49.130.3
WARN  2023-07-12 23:02:54,868 [shard 0] gossip - unknown endpoint 127.49.130.2
WARN  2023-07-12 23:02:54,869 [shard 0] gossip - unknown endpoint 127.49.130.3
WARN  2023-07-12 23:02:54,869 [shard 0] gossip - unknown endpoint 127.49.130.2
```

Reloading of topology from system.topology calls `storage_service::update_peer_info()`, which doesn't insert endpoint state:

```
future<> storage_service::update_peer_info(gms::inet_address endpoint) {
    slogger.debug("Update peer info: endpoint={}", endpoint);
    using namespace gms;
    auto* ep_state = _gossiper.get_endpoint_state_for_endpoint_ptr(endpoint);
    if (!ep_state) {
        co_return;
    }
```

The node doesn't see anyone to gossip with, so it won't learn about other nodes (on node .4):
```
TRACE 2023-07-12 23:01:53,364 [shard 0] gossip - === Gossip round START
TRACE 2023-07-12 23:01:53,364 [shard 0] gossip - My heartbeat is now 22
DEBUG 2023-07-12 23:01:53,364 [shard 0] gossip - Set live nodes to talk: endpoint_state_map=1, all_live_nodes={}, endpoints_to_talk_with={}
DEBUG 2023-07-12 23:01:53,364 [shard 0] gossip - No live nodes yet: try initial contact point nodes={}
DEBUG 2023-07-12 23:01:53,364 [shard 0] gossip - No one to talk with
TRACE 2023-07-12 23:01:53,364 [shard 0] gossip - Performing status check ...
DEBUG 2023-07-12 23:01:53,365 [shard 0] gossip - failure_detector_loop: Wait until live_nodes={} is not empty
TRACE 2023-07-12 23:01:53,365 [shard 0] gossip - === Gossip round OK
```

The node initially gossips only with seeds. If seed pointers form a graph which connects all the nodes, they all should eventually learn about each other. That's how seeds are configured here. For example, node .2 has seeds=[.4, .3]. However, so it happens that when .2 restarts, .4 is too early in the boot process and fails the gossip exchange:

```
WARN  2023-07-12 23:01:51,124 [shard 0] gossip - Fail to apply application_state: std::runtime_error (endpoint_state_map does not contain endpoint = 127.49.130.4, application_states = {{STATUS, Value(NORMAL,869247220184916093,1)}})
```

Only exchange with .3  completes. Then later, gossip will use only that (live) node, and won't use the other seed:

```
DEBUG 2023-07-12 23:01:50,968 [shard 0] gossip - Talk to live nodes: {127.49.130.3, 127.49.130.4}
....
TRACE 2023-07-12 23:01:51,974 [shard 0] gossip - === Gossip round START
TRACE 2023-07-12 23:01:51,974 [shard 0] gossip - My heartbeat is now 27
DEBUG 2023-07-12 23:01:51,974 [shard 0] gossip - Talk to live nodes: {127.49.130.3}
TRACE 2023-07-12 23:01:51,974 [shard 0] gossip - Sending a GossipDigestSyn to 127.49.130.3:0 ...
TRACE 2023-07-12 23:01:51,975 [shard 0] gossip - Performing status check ...
DEBUG 2023-07-12 23:01:51,983 [shard 0] gossip - replicating live endpoints to other shards
TRACE 2023-07-12 23:01:51,986 [shard 0] gossip - === Gossip round OK
```

In general, it is not guaranteed that seeds will be up to date and reachable at the time of restart, so we shouldn't rely on that.

--- a/gms/gossiper.cc
+++ b/gms/gossiper.cc
@@ -340,7 +340,18 @@ future<> gossiper::do_send_ack2_msg(msg_addr from, utils::chunked_vector<gossip_
         std::map<inet_address, endpoint_state> delta_ep_state_map;
         for (auto g_digest : ack_msg_digest) {
             inet_address addr = g_digest.get_endpoint();
-            auto local_ep_state_ptr = this->get_state_for_version_bigger_than(addr, version_type(g_digest.get_max_version()));
+            const auto es = get_endpoint_state_for_endpoint_ptr(addr);
+            if (!es || es->get_heart_beat_state().get_generation() < g_digest.get_generation()) {
+                continue;
+            }
+            // Local generation for addr may have been increased since the
+            // current node sent an initial SYN. Comparing versions across
+            // different generations in get_state_for_version_bigger_than
+            // could result in loosing some app states with smaller versions.
+            const auto version = es->get_heart_beat_state().get_generation() > g_digest.get_generation()
+                ? version_type(0)
+                : g_digest.get_max_version();
+            auto local_ep_state_ptr = this->get_state_for_version_bigger_than(addr, version);
             if (local_ep_state_ptr) {
                 delta_ep_state_map.emplace(addr, *local_ep_state_ptr);
             }

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -459,6 +459,18 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
         }
     }));
 
+    // We don't load gossiper endpoint states in storage_service::join_cluster
+    // if _raft_topology_change_enabled. On the other hand gossiper is still needed
+    // even in case of _raft_topology_change_enabled mode, since it still contains part
+    // of the cluster state. To work correctly, the gossiper needs to know the current
+    // endpoints. We cannot rely on seeds alone, since it is not guaranteed that seeds
+    // will be up to date and reachable at the time of restart.
+    for (const auto& e: get_token_metadata_ptr()->get_all_endpoints()) {
+        if (!utils::fb_utilities::is_me(e) && !_gossiper.get_endpoint_state_for_endpoint_ptr(e)) {
+            co_await _gossiper.add_saved_endpoint(e);
+        }
+    }
+
     if (auto gen_id = _topology_state_machine._topology.current_cdc_generation_id) {
         slogger.debug("topology_state_load: current CDC generation ID: {}", *gen_id);
         co_await cdc_gen_svc.handle_cdc_generation(*gen_id);

